Conversion of Lua PROTO files to JavaScript

ad-daniel opened the pull request:
**Description**
This PR contains the conversion of all PROTO files from Lua to JavaScript. When the two PROTO are somewhat similar, the review is fairly easy but  for those where the two versions are significantly different it might be easier to open them in dummy world and play with them to see if something breaks.

For the `WbQjsFile` module, the functions I exposed were the minimal I needed for the conversion. So, I don't know if this module should be more complex than that (or something more generic). Since removing/changing these functions becomes harder after publishing (for retro-compatibility reasons), it's worth deciding now.

Mostly complex/long ones & a couple simple ones that are important:

1. projects/objects/road/protos/AddLanesRoadSegment.proto
2. projects/objects/geometries/protos/TexturedParallelepiped.proto
3. projects/samples/contests/robocup/protos/RobocupGoal.proto
4. projects/objects/trees/protos/Forest.proto
5. projects/samples/robotbenchmark/pit_escape/protos/Pit.proto
6. projects/objects/floors/protos/CircleArena.proto
7. projects/objects/road/protos/Road.proto
8. projects/default/protos/SumoInterface.proto
9. projects/objects/backgrounds/protos/TexturedBackgroundLight.proto
10. projects/objects/buildings/protos/StripBuilding.proto
11. projects/objects/buildings/protos/FastFoodRestaurant.proto
12. projects/objects/road/protos/AddLaneRoadSegment.proto
13. projects/vehicles/protos/abstract/Car.proto
14. projects/vehicles/protos/abstract/VehicleWheel.proto
15. projects/objects/mirror/protos/Mirror.proto
16. projects/objects/road/protos/Roundabout.proto
17. projects/objects/street_furniture/protos/Fence.proto
18. projects/objects/road/protos/RoadIntersection.proto
19. projects/objects/geometries/protos/Extrusion.proto
20. projects/robots/gctronic/e-puck/protos/E-puck.proto
21. projects/objects/solids/protos/SolidRoundedBox.proto
22. projects/objects/floors/protos/UnevenTerrain.proto
23. projects/objects/factory/conveyors/protos/ConveyorBelt.proto
24. projects/robots/robotis/darwin-op/protos/Darwin-op.proto
25. projects/objects/gear/protos/Gear.proto
26. projects/objects/solids/protos/SolidPipe.proto
27. projects/objects/solids/protos/SolidBox.proto
28. projects/bounding_objects/protos/PipeBoundingObject.proto <--- crashes master/develop if change params, loading from file works
29. projects/joints/protos/HingeJointWithBacklash.proto
30. projects/objects/buildings/protos/SimpleBuilding.proto
31. projects/objects/buildings/protos/Building.proto
32. projects/joints/protos/Hinge2JointWithBacklash.proto
33. projects/objects/stairs/protos/StraightStairs.proto
34. projects/vehicles/protos/abstract/AckermannVehicle.proto
35. projects/objects/shapes/protos/TexturedBoxShape.proto
36. projects/objects/floors/protos/RectangleArena.proto
37. projects/samples/robotbenchmark/wall_following/protos/LinkedWalls.proto
38. projects/objects/trees/protos/SimpleTree.proto
39. projects/objects/panels/protos/Panel.proto
40. projects/samples/contests/robocup/protos/RobocupSoccerField.proto
41. projects/robots/robotis/darwin-op/protos/Darwin-opHinge2.proto
42. projects/objects/factory/pallet/protos/WoodenPalletStack.proto
43. projects/objects/advertising_board/protos/AdvertisingBoard.proto
44. projects/robots/softbank/nao/protos/Nao.proto
45. projects/objects/cabinet/protos/Cabinet.proto
46. projects/objects/road/protos/CurvedRoadSegment.proto
47. projects/objects/road/protos/LaneSeparation.proto
48. projects/objects/road/protos/HelicoidalRoadSegment.proto
49. projects/objects/factory/pallet/protos/WoodenPallet.proto
50. projects/objects/road/protos/CrashBarrier.proto <-- if sloped path, extrusion flips: already broken in develop/master

and non-proto files.

BenjaminDeleze commented:
As it is a quite heavy PR, I propose that we split the reviewing work as follow:
@omichel could you be the main reviewer as I think you followed this PR ?
And we will help you to review the protos listed in the description of this PR that need deep check:

- I will review the 10th first: from `projects/objects/road/protos/AddLanesRoadSegment.proto` to `projects/objects/buildings/protos/StripBuilding.proto`
- @lukicdarkoo will review from 11 to 20: from `projects/objects/buildings/protos/FastFoodRestaurant.proto` to `projects/robots/gctronic/e-puck/protos/E-puck.proto`
- @mael25 will review from 21 to 30: from `projects/objects/solids/protos/SolidRoundedBox.proto` to `projects/objects/buildings/protos/SimpleBuilding.proto`
- @stefaniapedrazzi will review from 31 to 40: from `projects/objects/buildings/protos/Building.proto` to `projects/objects/panels/protos/Panel.proto`
- and @omichel will take from 41 to 50: from `projects/samples/contests/robocup/protos/RobocupSoccerField.proto` to `projects/objects/factory/pallet/protos/WoodenPallet.proto`

lukicdarkoo commented:
**AddLaneRoadSegment**
-  The new PROTO generates a full line instead of dashed
  ![image](https://user-images.githubusercontent.com/2135826/122183679-c40e9c80-ce8b-11eb-85d8-0a2c2172658f.png)
- Deleting the RoadLane crashes Webots (not regression, we can make an issue about that)

**Fence**
- There is an extra pole:
  ![image](https://user-images.githubusercontent.com/2135826/122187309-3765dd80-ce8f-11eb-8cfb-ef6afce69098.png)

mael25 commented on projects/objects/solids/protos/SolidRoundedBox.proto:
```diff
@@ -17,258 +18,255 @@ PROTO SolidRoundedBox [
-    innerSize = { x = size.x - borderDiameter; y = size.y - borderDiameter; z = size.z - borderDiameter }
-  }%
+    if (2 * borderRadius > size.x || 2 * borderRadius > size.y || 2 * borderRadius > size.z) {
+      borderSize = 0.5 * Math.min(size.x, size.y, size.z);
```
This creates an error as `borderSize` was not defined. This was already the case in the Lua version but did not raise an error as there is no control of that in Lua. I think it was only a typo in the first place and that you can replace `borderSize` by `borderRadius` (at l.45 and l.46).

Also, the error message (l.46) and the test effectuated (l.44) differs and by testing it it seems that the error should occurs when 2*`borderRadius` >= min(size) instead of border 2*`borderRadius` > min(size). Then, `borderRadius` should be reset to a value slightly smaller than min(size)/2. Otherwise, the SolidRoundedBox looks weird. Maybe this second issue can be addressed in another PR as it is a bug present already in the Lua version.

stefaniapedrazzi commented on projects/samples/contests/robocup/protos/RobocupSoccerField.proto:
```diff
@@ -11,282 +12,267 @@ PROTO RobocupSoccerField [
+    // There is a vertex junction_index of the inner circle (cin) that makes the junction between triangles
+    // going through pt.40 (top vertex of central dash) and pt.35 (bottom vertex of central dash).
+    // It is located at approximately two thirds of the quarter
+    let junction_index = Math.ceil((2 / 3) * n_vertices_quarter); // Index of the junction triangle within the inner circle
```
This can be const
```suggestion
    const junction_index = Math.ceil((2 / 3) * n_vertices_quarter); // Index of the junction triangle within the inner circle
```

stefaniapedrazzi commented on projects/samples/contests/robocup/protos/RobocupSoccerField.proto:
```diff
@@ -11,282 +12,267 @@ PROTO RobocupSoccerField [
-      tl[3 * Ntl + i] = {a = tl_ref.c + 3 * Npts, b = tl_ref.a + 3 * Npts, c = tl_ref.b + 3 * Npts}
-    end
+    // Replication of the other 3 quarters
+    Npts = pts.length;
```
it seems that this can be remove
```suggestion
```

stefaniapedrazzi commented on projects/samples/contests/robocup/protos/RobocupSoccerField.proto:
```diff
@@ -11,282 +12,267 @@ PROTO RobocupSoccerField [
+    // Replication of the other 3 quarters
+    Npts = pts.length;
+    for (let i = 0; i < Npts; ++i) {
+      let pt_ref = pts[i]; // Reference point
```
This can be const
```suggestion
      const pt_ref = pts[i]; // Reference point
```

stefaniapedrazzi commented on projects/samples/contests/robocup/protos/RobocupSoccerField.proto:
```diff
@@ -11,282 +12,267 @@ PROTO RobocupSoccerField [
-      pts[i].x = pts[i].x - OFFSET_X
-      pts[i].y = pts[i].y - OFFSET_Y
-    end
+    let Ntg = tg.length;
+    for (let i = 0; i < Ntg; ++i) { // Grass triangles
+      let tg_ref = tg[i];
```
These can be const
```suggestion
    const Ntg = tg.length;
    for (let i = 0; i < Ntg; ++i) { // Grass triangles
      const tg_ref = tg[i];
```

stefaniapedrazzi commented on projects/samples/contests/robocup/protos/RobocupSoccerField.proto:
```diff
@@ -11,282 +12,267 @@ PROTO RobocupSoccerField [
+    }
 
-  }%
+    let Ntl = tl.length;
+    for (let i = 0; i < Ntl; ++i) { // Same for the line triangles
+      let tl_ref = tl[i];
```
These can be const
```suggestion
    const Ntl = tl.length;
    for (let i = 0; i < Ntl; ++i) { // Same for the line triangles
      const tl_ref = tl[i];
```

stefaniapedrazzi commented on projects/samples/contests/robocup/protos/RobocupSoccerField.proto:
```diff
@@ -11,282 +12,267 @@ PROTO RobocupSoccerField [
+    }
 
+    // Offset to have origin on the center of the field
+    let OFFSET_X = (2 * I + A) / 2;
+    let OFFSET_Y = (2 * I + B) / 2;
```
These can be const
```suggestion
    const OFFSET_X = (2 * I + A) / 2;
    const OFFSET_Y = (2 * I + B) / 2;
```

stefaniapedrazzi reviewed (changes requested):
I checked the PROTO files assigned to me.
There some minor code improvement but all works correctly except:
* CrashBarrier PROTO: set subdivion to 3 and enableBoundingObject to TRUE
    ```
    ERROR: '/home/stefi/webots/projects/objects/road/protos/CrashBarrier.proto': error: Template engine error: failed to execute JavaScript template: Cannot read property 'x' of undefined.
    ERROR: Template regeneration failed. The node cannot be generated.
    ```
    Spline is working on the develop branch but not in this branch, for example when setting the path to [0 0 0, 0 0 2, 5 0 0].
    Instead of a spline, in this branch the result are straight segments.

omichel commented on projects/objects/road/protos/HelicoidalRoadSegment.proto:
```diff
@@ -79,14 +80,14 @@ PROTO HelicoidalRoadSegment [
+        const turnNumber = height / heigthStep;
+        const heightSubdivisionStep = height / subdivision;
+        const angleSubdivisionStep = (Math.PI * turnNumber) / subdivision;
+      >%
+      %< for (let i = 0; i <= subdivision; ++i) { >%
```
```suggestion
        for (let i = 0; i <= subdivision; ++i) { >%
```

omichel commented on projects/objects/factory/pallet/protos/WoodenPalletStack.proto:
```diff
@@ -19,89 +20,88 @@ PROTO WoodenPalletStack [
-  end
+  let palletSize = fields.palletSize.value;
+  if (palletSize.x <= 0 || palletSize.y <= 0 || palletSize.z <= 0) {
+    size = fields.size.defaultValue;
```
```suggestion
    palletSize = fields.palletSize.defaultValue;
```

omichel commented on projects/objects/factory/pallet/protos/WoodenPalletStack.proto:
```diff
@@ -19,89 +20,88 @@ PROTO WoodenPalletStack [
+  let palletSize = fields.palletSize.value;
+  if (palletSize.x <= 0 || palletSize.y <= 0 || palletSize.z <= 0) {
+    size = fields.size.defaultValue;
+    console.error('\'size\' should be strictly positive. Value reset to (' + size.x + ', ' + size.y + ', ' + size.z + ').');
```
```suggestion
    console.error('\'palletSize\' should be strictly positive. Value reset to (' + palletSize.x + ', ' + palletSize.y + ', ' + palletSize.z + ').');
```

omichel commented on projects/objects/advertising_board/protos/AdvertisingBoard.proto:
```diff
@@ -118,53 +103,53 @@ PROTO AdvertisingBoard [
-          }%
+          %<
+            if (fields.frontTexture.value.length === 0) {
+              // get size of Lua table component to decide whether an url is present or not.
```
```suggestion
              // get length to decide whether an url is present or not.
```

mael25 commented:
In `UnevenTerrain`: error if perlinNOctaves >=10 (also in LUA)
ERROR: '/home/mael/webots/projects/objects/floors/protos/UnevenTerrain.proto': error: Template engine error: failed to execute JavaScript template: Cannot read property '0' of undefined.

In `Gear`: negative value are possible for `innerRadius` and `toothDepth` (also in LUA)

mael25 commented on projects/bounding_objects/protos/PipeBoundingObject.proto:
```diff
@@ -11,79 +12,79 @@ PROTO PipeBoundingObject [
+    const alpha = beta * 0.5;
+    const innerRadius = radius - thickness;
+    const su = radius * Math.cos(alpha) - innerRadius;
+    if (su < 0) {
```
If su < 0 there is a crash and we don't see your console message:

ERROR: '/home/mael/webots/projects/bounding_objects/protos/PipeBoundingObject.proto': error: Template engine error: failed to execute JavaScript template: .

(no crash in LUA)

Whereas in LUA, there

lukicdarkoo commented on projects/objects/road/protos/Roundabout.proto:
```diff
@@ -150,60 +148,59 @@ PROTO Roundabout [
+          let centralIndex = 0;
+          let pointIndex = 0;
+          let pointsNumber = outerPoints.length;
+          for (let i = 0; i < pointsNumber; ++i) {
+            if (wbvector2.equal(outerPoints[i], point)) {
+              pointIndex = i;
+            } else if (wbvector2.equal(outerPoints[i], centralPoint)) {
+              centralIndex = i;
+            }
+          }
```
```suggestion
          for (let i = 0; i < pointsNumber; ++i)
            if (wbvector2.equal(outerPoints[i], point))
              pointIndex = i;
            else if (wbvector2.equal(outerPoints[i], centralPoint))
              centralIndex = i;
```
Cyberbotics rule is to avoid brackets

lukicdarkoo commented on projects/objects/road/protos/Roundabout.proto:
```diff
@@ -150,60 +148,59 @@ PROTO Roundabout [
-          width %{= roadsWidth }%
+          translation %<= centralPoint.x >% 0 %<= centralPoint.y >%
+          rotation 0 1 0 %<= rotationAngle >%
+          name "road border %<=i>%"
+          width %<= roadsWidth >%
```
```suggestion
          name "road border %<= i >%"
          width %<= roadsWidth >%
```
Spacing is not consistent, I would rather use spaces before and after the expression

ad-daniel replied:
This is a special case as the evaluation is inside of a string, that's why there are no spaces or they'd show up in the name as well. Either way I changed it.

lukicdarkoo commented on projects/objects/street_furniture/protos/Fence.proto:
```diff
@@ -28,102 +29,98 @@ PROTO Fence [
+      %< if (fields.poleAppearance.value !== undefined) { >%
+        %< if (fields.poleAppearance.value.fields.baseColor !== undefined) { >%
+          %< let color = fields.poleAppearance.value.fields.baseColor.value; >%
+          %<= color.r >% %<= color.g >% %<= color.b >%
```
```suggestion
          %<= `${color.r} ${color.g} ${color.b}` >%
```
One more place in which the templates may help clarity

ad-daniel commented on projects/objects/panels/protos/Panel.proto:
```diff
@@ -15,129 +16,126 @@ PROTO Panel [
+                  if (nSeparations > 0) {
+                    let offsetX = 1.5 * borderHeight + panelSizeX; // halfSizeX
+                    for (let i = 1; i <= nSeparations; ++i) {
+                      for (let dz = -0.5; dz <= 0.5; dz = dz + 1) {
```
no logical reason

ad-daniel commented:
General remarks:
- The crashes (if they weren't regressions) have not been fixed in this PR as I don't think it's related to the PROTO themselves. I'll open an issue about all of those.
- For all roads, field `dashedLane` was deprecated in 2018 and was making it a living hell to fix the pre-existing Lua bugs while ensuring that this field continued to worked as before. So I removed it and adapted the worlds that specified it accordingly.
- Concerning the `const`, the reason I was parsimonious on those is because for whatever reason, if you attempt to change a `const` variable, this engine just fails without telling you why. So since it was already fairly easy to mess up the conversion itself, I've put it only where I was fairly certain I wouldn't fall into this problem. However now that they load, it's true that more `const` can be added so I've gone through these 50 already. The remaining 200+ PROTO might benefit from it too, but I'd rather do it as a second round of cleanup as this PR is already unmanageable. Also fixed the way undefined was checked as per the coding style and unnecessary brackets.
   - [From 0 to 10](https://github.com/cyberbotics/webots/pull/3156/commits/c29ad7131085cf01904747d29bb3f30d6f2f4a22) 
   - [From 10 to 20](https://github.com/cyberbotics/webots/pull/3156/commits/f25226c924957cb2137c3e7baed1e304a4180869)
   - [From 20 to 30](https://github.com/cyberbotics/webots/pull/3156/commits/7eda7728a3369b19d855dac948bd985c6a7eeeec)
   - [From 30 to 40](https://github.com/cyberbotics/webots/pull/3156/commits/e8c0902bafe704f5bb1b78ba384faec191b76530)
   - [From 40 to 50](https://github.com/cyberbotics/webots/pull/3156/commits/27f91aa6b6a0b9fddc47cc7940a09e1f2db60ddf)
- For template literals, check Darko's comments.

@lukicdarkoo:
- Fence: fixed
- AddLaneRoadSegment: fixed. As you might notice, more changes have been made in this PROTO, and the reason is that another PROTO depends on it and in order to fix that one, changes had to be done here as well (the line of the added road for instance couldn't be of whatever type, but I need it to be)
- Concerning the `template literals`, I do agree that it looks a lot better and it might even make the loading faster as it involves less template evaluations. I made a few tests and it appears the engine support them. However this PR is already unmanageable, and since the other 200 PROTO would profit from it as well, I'd rather do it in a different PR. If that's ok with you I'll first make a test to see if indeed it's faster, and if so I'll optimize them later.

@mael25:
- Gear: fixed.
- UnevenTerrain: seems like a limit of the algorithm itself and isn't a regression, so I just added a check so that only `perlinNOctaves < 10` are accepted.
- SolidRoundedBox: initially I fixed it as you suggested, but as many worlds appears to do it already, I ended up changing the message instead. The reason it looks odd is because the size is small. The worlds that set it at that limit value, usually they're very flat (large surface, small height) and it makes sense there for it to be half the size (thickness).
- PipeBoundingObject: fixed crash (only the regression one).

@stefaniapedrazzi :
- Also fixed issue #3185, the fix is [here](https://github.com/cyberbotics/webots/pull/3156/commits/7d2d0ee97e0afae8d5a552f1fafa04aa2d5ab37f)
- CrashBarrier: fixed the crash. As for the spline comment, unfortunately I can't reproduce it.
![crashbarrier](https://user-images.githubusercontent.com/44834743/123065909-e23a4680-d40f-11eb-987c-03e245f3d360.png)

@omichel :
- AdvertisingBoard: fixed H-Shape
- `tab-components`: addressed in #3210
- WoodenPallet: `lathNumber` now coincides with visible laths.
- Also fixed issue #3184, the fix is [here](https://github.com/cyberbotics/webots/pull/3156/commits/8c668de4fc96a1fbac498c7df5fcc9d4ae6d5f5f)

@BenjaminDeleze :
- CircleArena: fixed, apparently the entire circle was provided as a single polygon to the `IndexedFaceSet` and for certain values it was unable to split it into triangles by itself. For the boundingObject, I fixed the fact that for certain subdivisions it doesn't match with the physical border. I think that's what you meant.
- Road: I don't think I can reproduce it:
![road](https://user-images.githubusercontent.com/44834743/123068878-950ba400-d412-11eb-8e48-1928ce84240a.png)

I might have fixed it unintentionally in the process though.
- AddLanesRoadSegment: a nightmare but right now it works and in theory it supports all types of lines which previously did not.
- Forest: I couldn't find any depth test being made. If you mean that when you look at the forest sideways it appears that some trees are invisible (and that you can see some trees that you shouldn't) then yes, that might be a visual bug but I'd assume is on the webots side, not PROTO.

Hopefully I didn't forget anything. Let me know if I did.

Note: after merging this PR, viewer.js must be uploaded to the server.

omichel reviewed (dismissed):
All good for me.
Thank you!

lukicdarkoo reviewed (dismissed):
> If that's ok with you I'll first make a test to see if indeed it's faster, and if so I'll optimize them later.

Sounds good to me. Thank you

stefaniapedrazzi reviewed (changes requested):
Thank you. All the changes are working correctly.
I checked and also the spline issue with the Crashbarrier is fixed now (I can still reproduce it checking out the version I reviewed last time).

Just a minor issue in an error message due to my previous suggestion.

stefaniapedrazzi commented on projects/objects/trees/protos/SimpleTree.proto:
```diff
@@ -31,68 +32,58 @@ PROTO SimpleTree [
+    let candidates = '';
+    for (let i = 0; i < keys.length; ++i)
+      candidates += '\'' + keys[i] + '\', ';
+    candidates += '\' or \'random\'';
```
Sorry. my previous suggestion was wrong
```suggestion
    candidates += 'or \'random\'';
```

ad-daniel replied:
I missed it too, thank you

stefaniapedrazzi reviewed (approved):
Thank you!

BenjaminDeleze reviewed (dismissed):
Thank you!

Everything seems good for me.

mael25 reviewed (approved):
Thank you, everything works well now!
